turbine/src/xdp.rs

```diff
@@ -0,0 +1,183 @@
+    pub cpus: Vec<usize>,
+    pub zero_copy: bool,
+    pub channel_cap: usize,
+}
+
+impl XdpConfig {
+    // A nice round number
+    const DEFAULT_CHANNEL_CAP: usize = 1_000_000;
+}
+
+impl Default for XdpConfig {
+    fn default() -> Self {
+        Self {
+            interface: None,
+            cpus: vec![],
+            zero_copy: false,
+            channel_cap: Self::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+impl XdpConfig {
+    pub fn new(interface: Option<impl Into<String>>, cpus: Vec<usize>, zero_copy: bool) -> Self {
+        Self {
+            interface: interface.map(|s| s.into()),
+            cpus,
+            zero_copy,
+            channel_cap: XdpConfig::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        shred_index: u32,
+        addr: Vec<SocketAddr>,
+        payload: shred::Payload,
+    ) -> Result<(), TrySendError<(Vec<SocketAddr>, shred::Payload)>> {
+        self.senders[shred_index as usize % self.senders.len()].try_send((addr.clone(), payload))
+    }
+}
+
+pub(crate) struct XdpRetransmitter {
+    threads: Vec<thread::JoinHandle<()>>,
+}
+
+impl XdpRetransmitter {
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn new(
+        _config: XdpConfig,
+        _src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        Err("XDP is only supported on Linux".into())
+    }
+
+    #[cfg(target_os = "linux")]
+    pub(crate) fn new(
+        config: XdpConfig,
+        src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        use caps::{
+            CapSet,
+            Capability::{CAP_BPF, CAP_NET_ADMIN, CAP_NET_RAW},
+        };
+
+        // switch to higher caps while we setup XDP. We assume that an error in
+        // this function is irrecoverable so we don't try to drop on errors.
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::raise(None, CapSet::Effective, cap)
+                .map_err(|e| format!("failed to raise {cap:?} capability: {e}"))?;
+        }
+
+        let dev = Arc::new(if let Some(interface) = config.interface {
+            NetworkDevice::new(interface).unwrap()
+        } else {
+            NetworkDevice::new_from_default_route().unwrap()
+        });
+
+        let ebpf = if config.zero_copy {
+            Some(
+                load_xdp_program(dev.if_index())
+                    .map_err(|e| format!("failed to attach xdp program: {e}"))?,
+            )
+        } else {
+            None
+        };
+
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::drop(None, CapSet::Effective, cap).unwrap();
+        }
+
+        let (senders, receivers) = (0..config.cpus.len())
+            .map(|_| crossbeam_channel::bounded(config.channel_cap))
+            .unzip::<_, _, Vec<_>, Vec<_>>();
```

Comment:
if channels are per-cpu, creating multiple channels should share the buffer capacity no? Else setting high cpu count may have undesired effect of massively increasing max RAM usage.

Reply:
a review comment phrased as a question!! I must be dreaming!! 😋

Maybe? If the NIC supports XDP natively, there's no reason to use more than once CPU. When testing XDP in fallback SKB mode, each core can do 7-10Gbps, so one might need a couple of CPUs. 

"Massively" is an overstatement. It's 2x channel cap, compared to... a whole number of unbounded channel along the way to this channel.

Reply:
Are you dreaming? How would you even know if you are? If you are, can you imagine all the obnoxious settings people will try to feed to agave in prod? If the buffer size is governed by the bursty nature of the retransmit stage packet production, would it not make sense to decouple the buffer sizes from the # of cores used (irrsepective of how many are needed)? Or, alternatively, maybe we should hardwire the number of cores to 1 if that is sufficient in all cases and avoid the extra complexity?

Reply:
You are assuming that load balancing from the code that sends into these channels is perfectly fair, which is not, and you wouldn't want it to be. So since it's not fair, how do we split the capacity? What's the right formula? Does it make sense to fine tune this at this stage of this whole XDP development?

The answer seems _obviously_ no.

Reply:
In fact that's exactly what I did in an earlier iteration, and then I took it out to simplify the code, considering that it can only be triggered in extreme conditions that don't happen on mnb. 

IMO there's zero point fine tuning this down. I would like to merge this PR once real issues are addressed, then fix the bursty behavior, then after that come back and fine tune given the numbers we get once the bursts are fixed.

Reply:
fair point, let us then add monitoring of some kind such that we know if serious imbalance is present? e.g. check every second all buffer occupancies and report mean and highest to metrics?

Reply:
punting on this - as explained all channel handling in turbine needs to be reworked. I'll do it as part of that.